# Introduction to Data Science 2022

# Week 1

## Exercise 1 | Matrix warm-up
<span style="background-color: #ccfff2"> *Note: You can find tutorials for NumPy and Pandas under 'Useful tutorials' on Moodle.*</span>

One of the most useful properties of any scientific programming language (Python with NumPy, R, Julia, etc) is that they allow us to work with matrices efficiently. Let's learn more about these features!

### 1.1 Basics

1. Let's start by creating two arrays <span style="background-color: #ccfff2"> A</span> and <span style="background-color: #ccfff2"> B</span> which each have the integers <span style="background-color: #ccfff2"> 0, 1, 2, ..., 1e7-1</span>. Use the normal arrays or lists of the programming language you are using, e.g. *list* or *[ ]* or *numpy.array()* in Python.

In [2]:
import numpy as np

A = np.arange(0, 1e7 - 1, dtype= int)
B = np.arange(0, 1e7 - 1, dtype= int)

2. Create a function that uses a <span style="background-color: #ccfff2"> for loop</span> or equivalent to return a new array <span style="background-color: #ccfff2"> C</span>, which contains the <span style="background-color: #ccfff2"> element-wise sum of *A* and *B*</span>, e.g. C should contain the integers <span style="background-color: #ccfff2"> 0, 2, 4, etc</span>.

In [3]:
def add_with_for(A,B):
    C = []
    # add your code here
    for i in range(min(len(A), len(B))):
        C.append(A[i] + B[i])
    return C

add_with_for(A,B)

[0,
 2,
 4,
 6,
 8,
 10,
 12,
 14,
 16,
 18,
 20,
 22,
 24,
 26,
 28,
 30,
 32,
 34,
 36,
 38,
 40,
 42,
 44,
 46,
 48,
 50,
 52,
 54,
 56,
 58,
 60,
 62,
 64,
 66,
 68,
 70,
 72,
 74,
 76,
 78,
 80,
 82,
 84,
 86,
 88,
 90,
 92,
 94,
 96,
 98,
 100,
 102,
 104,
 106,
 108,
 110,
 112,
 114,
 116,
 118,
 120,
 122,
 124,
 126,
 128,
 130,
 132,
 134,
 136,
 138,
 140,
 142,
 144,
 146,
 148,
 150,
 152,
 154,
 156,
 158,
 160,
 162,
 164,
 166,
 168,
 170,
 172,
 174,
 176,
 178,
 180,
 182,
 184,
 186,
 188,
 190,
 192,
 194,
 196,
 198,
 200,
 202,
 204,
 206,
 208,
 210,
 212,
 214,
 216,
 218,
 220,
 222,
 224,
 226,
 228,
 230,
 232,
 234,
 236,
 238,
 240,
 242,
 244,
 246,
 248,
 250,
 252,
 254,
 256,
 258,
 260,
 262,
 264,
 266,
 268,
 270,
 272,
 274,
 276,
 278,
 280,
 282,
 284,
 286,
 288,
 290,
 292,
 294,
 296,
 298,
 300,
 302,
 304,
 306,
 308,
 310,
 312,
 314,
 316,
 318,
 320,
 322,
 324,
 326,
 328,
 330,
 332,
 334,
 336,
 338,
 340,
 342,
 344,
 346,
 348,
 350,

3. Next, let's create another function that uses NumPy (or equivalent) to do the same. To try it out, allocate two arrays (e.g. using <span style="background-color: #ccfff2"> np.array</span> in NumPy) and add the arrays together using your function. Don't use loops, instead, find out how to add the two arrays directly. What do you notice in comparison to the previous function?

In [4]:
def add_without_loop(A,B):
    C = []
    C.append(A+B)
    return C

add_without_loop(A,B)

[array([       0,        2,        4, ..., 19999992, 19999994, 19999996])]

### 1.2 Array manipulation

<span style="background-color: #ccfff2"> *Note: for the following exercises, only use NumPy or equivalent functions. Don't use any loops.* </span>
1. Create the following array:

*[hint: <span style="background-color: #ccfff2"> np.reshape</span>]*

In [5]:
# array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
#        [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
#        [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
#        [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
#        [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
#        [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
#        [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
#        [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
#        [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
#        [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]])

In [6]:
# Use this cell for your code
A = np.arange(100)
np.reshape(A, (10,10))


array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
       [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
       [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
       [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
       [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
       [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
       [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
       [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]])

2. Create the following array:

In [7]:
# array([[0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
#        [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.]])

In [8]:
# Use this cell for your code
y = np.zeros([10,10] , dtype=float)
np.full_like(y, [0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0])

array([[0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
       [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
       [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
       [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
       [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
       [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
       [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
       [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
       [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.],
       [0., 1., 0., 1., 0., 1., 0., 1., 0., 1.]])

3. Create the following array (D):

In [9]:
# array([[0., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 0., 1., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 0., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 0., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 0., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 0., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 0., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 1., 0., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 1., 1., 0.]])

In [10]:
# Use this cell for your code
#pattern

D = np.ones([10,10], dtype = float)

np.fill_diagonal(D, 0)

D



array([[0., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 0., 1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 0., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 0., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 0., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 0., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 0., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 0.]])

4. Create the following array (E):

In [11]:
# array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
#        [1., 1., 1., 1., 1., 1., 1., 1., 0., 1.],
#        [1., 1., 1., 1., 1., 1., 1., 0., 1., 1.],
#        [1., 1., 1., 1., 1., 1., 0., 1., 1., 1.],
#        [1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
#        [1., 1., 1., 1., 0., 1., 1., 1., 1., 1.],
#        [1., 1., 1., 0., 1., 1., 1., 1., 1., 1.],
#        [1., 1., 0., 1., 1., 1., 1., 1., 1., 1.],
#        [1., 0., 1., 1., 1., 1., 1., 1., 1., 1.],
#        [0., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

In [12]:
# Use this cell for your code
E = np.ones([10,10], dtype = float)

np.fill_diagonal(np.fliplr(E), 0)

E



array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 0.],
       [1., 1., 1., 1., 1., 1., 1., 1., 0., 1.],
       [1., 1., 1., 1., 1., 1., 1., 0., 1., 1.],
       [1., 1., 1., 1., 1., 1., 0., 1., 1., 1.],
       [1., 1., 1., 1., 1., 0., 1., 1., 1., 1.],
       [1., 1., 1., 1., 0., 1., 1., 1., 1., 1.],
       [1., 1., 1., 0., 1., 1., 1., 1., 1., 1.],
       [1., 1., 0., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 1., 1., 1., 1., 1., 1., 1., 1.],
       [0., 1., 1., 1., 1., 1., 1., 1., 1., 1.]])

5. Call the last two matrices <span style="background-color: #ccfff2">D</span> and <span style="background-color: #ccfff2">E</span>, respectively. Show that the determinant of their product (matrix multiplication) is the same as the product of their determinants. That is calculate both <span style="background-color: #ccfff2">det(DE)</span> and <span style="background-color: #ccfff2">det(D) * det(E)</span>, and show that they are the same. Is it a coincidence? (I think not) The product of the determinants (or the determinant of the product) should be -81.

In [13]:
# Use this cell for your code
d = np.linalg.det(D)
e = np.linalg.det(E)
print(d * e)


-81.0


In [14]:
x = np.matmul(D,E)
print(np.linalg.det(x))

-81.00000000000003


-- *Use this markdown cell for your written answer* --

### 1.3 Slicing

Array slicing is a powerful way to extract data from an array. Let's practice array slicing with the following exercises!

1. Load the [California housing dataset](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_california_housing.html). The data should be a matrix of shape <span style="background-color: #ccfff2">(20640, 8)</span>, that is 20640 rows and 8 columns. Use the <span style="background-color: #ccfff2">.shape</span> attribute of NumPy arrays to verify this. Here's a [description of the fields](https://scikit-learn.org/stable/datasets/real_world.html#california-housing-dataset).

In [15]:
# Use this cell for your code
import sklearn
import numpy
from sklearn import datasets

arr = sklearn.datasets.fetch_california_housing(data_home=None, download_if_missing=True, return_X_y=False, as_frame=False)
print(arr.data.shape)


(20640, 8)


2. Select rows where the average number of bedrooms <span style="background-color: #ccfff2">(AveBedrms)</span> is higher than 2. The first few row indices should be <span style="background-color: #ccfff2">710,  1023,  1024, ...</span> (zero-indexed). Count these houses - how many rows are selected? *[hint: <span style="background-color: #ccfff2">np.where</span>]*

In [16]:
# Use this cell for your code
result = numpy.where(arr.data[:,3] > 2)
print(result)
print(result[0].shape[0])
#numpy.sum(result)

(array([  710,  1023,  1024,  1030,  1102,  1233,  1234,  1235,  1236,
        1238,  1239,  1240,  1566,  1867,  1871,  1872,  1879,  1881,
        1889,  1901,  1906,  1910,  1911,  1912,  1913,  1914,  1925,
        1926,  1978,  1979,  2392,  2395,  2396,  2397,  2398,  2510,
        2511,  2643,  2656,  2763,  2768,  2776,  2778,  3014,  3086,
        3095,  3258,  3292,  3293,  3313,  3314,  3334,  3349,  3350,
        9193,  9423,  9431,  9451,  9634,  9669,  9671,  9672,  9675,
        9676,  9678,  9679,  9680,  9681,  9682,  9683, 10067, 10076,
       10077, 10078, 10079, 10080, 10081, 10082, 10083, 10084, 10405,
       11705, 11706, 11707, 11708, 11709, 11710, 11711, 11713, 11714,
       11715, 11716, 11717, 11719, 11720, 11721, 11722, 11723, 11724,
       11725, 11729, 11831, 11832, 11833, 11834, 11847, 11848, 11849,
       11850, 11862, 11865, 11866, 11867, 11869, 11870, 12136, 12303,
       12305, 12306, 12307, 12324, 12325, 12344, 12349, 12351, 12354,
       12359, 12361

3. Select the rows where the median house age (i.e. median in each block group) <span style="background-color: #ccfff2">(HouseAge)</span> is between 1 and 3 years (inclusive). There should be **124** of these.

In [17]:
# Use this cell for your code
array = numpy.logical_and(arr.data[:,1] >= 1 , arr.data[:,1] <= 3)
#result = arr.data[numpy.where((arr.data[:,1] > 1) & (arr.data[:,1] < 3))]
result = numpy.where(array)
print(result)
#print(array)
print(result[0].shape[0])


(array([   59,   854,   871,   959,   976,  1260,  1565,  1566,  1648,
        1719,  2206,  2310,  2311,  2312,  2336,  2339,  2706,  2774,
        2993,  3130,  3140,  5506,  9109,  9132,  9147,  9168,  9187,
        9378,  9639, 10336, 10376, 10391, 10401, 10403, 10406, 10410,
       10412, 10416, 10440, 10456, 10467, 10513, 10514, 10519, 10521,
       10524, 10525, 10529, 10548, 10619, 10622, 10631, 11490, 11775,
       12000, 12006, 12077, 12084, 12119, 12123, 12137, 12142, 12143,
       12151, 12156, 12167, 12171, 12201, 12202, 12203, 12204, 12285,
       12286, 12308, 12363, 12430, 12841, 12868, 13041, 13074, 13100,
       13102, 13139, 13177, 13179, 13367, 13368, 13373, 13374, 13375,
       13422, 13659, 13713, 13715, 13747, 13848, 13855, 14557, 15220,
       15444, 15464, 15736, 15785, 16383, 16566, 16593, 17068, 17824,
       17826, 17827, 18725, 18942, 18947, 18972, 18989, 19001, 19003,
       19033, 19046, 19536, 19561, 19650, 20070, 20367]),)
124


4. Find the mean of the block group population <span style="background-color: #ccfff2">(Population)</span> for homes whose median value is more than 25000 USD (the target variable). It should be around **1425.68**.

In [18]:
# Use this cell for your code
A = numpy.where(arr.target > 0.25)
popl = arr.data[A , 4]
print(numpy.mean(popl))

1425.6833737275813


**Remember to submit your code on Moodle. You can return this Jupyter notebook (.ipynb) or .py, .R, etc depending on your programming preferences.**

## Exercise 2 | Working with text data

Next, let's look into some text data. We will be looking into Amazon reviews, and the necessary steps to transform a raw dataset into a format more suitable for prediction tasks.

1. Download the automotive 5-core dataset from [here](https://jmcauley.ucsd.edu/data/amazon_v2/categoryFilesSmall/Automotive_5.json.gz). Next, you can extract the data in <span style="background-color: #ccfff2">JSON</span> format. You can also download one of the bigger ones, if you are feeling ambitious. Open the JSON file. Access the <span style="background-color: #ccfff2">reviewText</span> field, which contains the unstructured review text written by the user.

For instance, the first review reads as follows: 

*'After I wrote the below review, the manufacturer contacted me and explained how to use this.  Instead of the (current) picture on Amazon where the phone is placed vertically, you actually use the stand with the phone placed horizontally. [...]'*

In [ ]:
# Use this cell for your code
import pandas as pd
data = pd.read_json('Automotive_5.json', lines = True)
#data.loc['reviewText']
text = data['reviewText']
print(text[0])


2. Next, let's follow some steps to normalize the text data.

When dealing with natural language, it is important to notice that while, for example, the words "Copper" and "copper" are represented by two different strings, they have the same meaning. When applying statistical methods on this data, it is useful to ensure that words with the same meaning are represented by the same string.

* <span style="background-color: #ccfff2">Downcasing</span>: Let's first downcase the contents of the <span style="background-color: #ccfff2">reviewText</span> field.

Now the first review should be:

*'after i wrote the below review, the manufacturer contacted me and explained how to use this.  instead of the (current) picture on amazon where the phone is placed vertically, you actually use the stand with the phone placed horizontally.'*

In [ ]:
# Use this cell for your code
text = text.str.lower()
print(text[0])

3. Let's continue with punctuation and stop word removal. Stop words are words like "and", "the", etc. They are usually very common words that have little to do with the actual content matter. There's plenty openly available lists of stop words for almost any (natural) language.

* <span style="background-color: #ccfff2">Punctuation and stop-word removal</span>: Let's now remove all punctuation, as well as the stop words. You can find a stop word list for English, e.g. [here](https://gist.github.com/xldrkp/4a3b1a33f10d37bedbe0068f2b4482e8#file-stopwords-en-txt).*(use the link to download a txt of english stopwords)* Save the stopwords in the file as "stopwords-en.txt".

First review at this point reads as: 

*'wrote review manufacturer contacted explained current picture amazon phone vertically stand phone horizontally'*

In [ ]:
import pandas as pd

textFile = pd.read_csv("stopwords-en.txt", sep=" ")  #to remove punctuation
pattern = r'[^\w\s]'
text = text.str.replace(pattern, '',regex= True)
data["reviewText"] = data['reviewText'].apply(lambda x: ' '.join([word for word in str(x).split() if word not in (stopwords)]))
print(text[0])

4. Let's continue with stemming. For example, while the words "swims" and "swim" are different strings, they both refer to swimming. [Stemming](https://en.wikipedia.org/wiki/Stemming) refers to the process of mapping words from their inflected form to their base form, for instance: swims -> swim.

* <span style="background-color: #ccfff2">Stemming</span>: Apply a stemmer on the paragraphs, so that inflected forms are mapped to the base form. For example, for Python the popular natural language toolkit [nltk](http://www.nltk.org/howto/stem.html) has an easy to use stemmer. In case you are using R, you can try the [Snowball stemmer](https://www.rdocumentation.org/packages/corpus/versions/0.10.2/topics/stem_snowball). You can find out how to install nltk from [here](https://www.nltk.org/install.html). It will take a while to run! So, grab a coffee and wait :D

Finally, after stemming: 

*'wrote review manufactur contact explain current pictur amazon phone vertic stand phone horizont'*

In [ ]:
# Use this cell for your code
from nltk.stem.porter import *

ps = PorterStemmer() # Takes a while
data["reviewText"] = data["reviewText"].apply(lambda x: ' '.join([ps.stem(word) for word in str(x).split()]))

print('Review example after stemming: ', data.iloc[1]['reviewText'])

5. Finally, filter the data by selecting reviews where the field <span style="background-color: #ccfff2">overall</span> is 4 or 5, and store the review texts in a file named <span style="background-color: #ccfff2">pos.txt</span>. Similarly, select reviews with rating 1 or 2 and store them in a file named <span style="background-color: #ccfff2">neg.txt</span>. Ignore the reviews with overall rating 3. Each line in the two files should contain exactly one preprocessed review text without the rating.

In [ ]:
#cannot understand the key error for 'reviewText'

import json 

data = []
positive = []
negative = []
for row in range(len(data)):
    if data.iloc[row]["overall"] > 3:
        positive.append(data.iloc[row]["reviewText"])
    elif data.iloc[row]["overall"] < 3:
        negative.append(data.iloc[row]["reviewText"])
        
with open("pos.txt", "w") as f:
    f.write("\n".join(positive))
    
with open("neg.txt", "w") as f:
    f.write("\n".join(negative))

Having created two collections of positive and negative reviews respectively, you may wish to take a quick look to see how the review texts differ between them (positive vs negative). Keep these files, as we will be using this data later to experiment with machine learning methods.

**Remember to submit your code on Moodle. You can return this Jupyter notebook (.ipynb) or .py, .R, etc depending on your programming preferences. No need to submit the text files.**